ratchet-tree.go

```diff
@@ -33,50 +34,48 @@ type LeafNodeHashInput struct {
 
 //// RatchetTree Node
 type RatchetTreeNode struct {
-	Cred           *Credential `tls:"optional"`
+	Credential     *Credential `tls:"optional"`
 	PublicKey      *HPKEPublicKey
 	PrivateKey     *HPKEPrivateKey
 	UnmergedLeaves []leafIndex `tls:"head=2"`
-	CipherSuite    CipherSuite
 }
 
-func newRatchetTreeNode(cs CipherSuite, secret []byte) *RatchetTreeNode {
-	if len(secret) > 0 {
-		priv, err := cs.hpke().Derive(secret)
-		if err != nil {
-			panic(fmt.Errorf("hpke private key generation failed: %v", err))
-		}
-
-		return &RatchetTreeNode{
-			Cred:           nil,
-			PrivateKey:     &priv,
-			PublicKey:      &priv.PublicKey,
-			UnmergedLeaves: []leafIndex{},
-			CipherSuite:    cs,
-		}
+// Compare the public aspects of two nodes
+func (n RatchetTreeNode) Equals(o RatchetTreeNode) bool {
+	lhsCredNil := (n.Credential == nil)
+	rhsCredNil := (o.Credential == nil)
+	if lhsCredNil != rhsCredNil {
+		return false
+	}
+	if !lhsCredNil && !reflect.DeepEqual(n.Credential, o.Credential) {
+		return false
 	}
 
-	return &RatchetTreeNode{
-		Cred:           nil,
-		PrivateKey:     nil,
-		PublicKey:      nil,
-		UnmergedLeaves: []leafIndex{},
-		CipherSuite:    cs,
+	if !reflect.DeepEqual(n.PublicKey, o.PublicKey) ||
+		!reflect.DeepEqual(n.UnmergedLeaves, o.UnmergedLeaves) {
```

Comment:
Note: This test will fail if one side is `nil` and the other is a zero-length array, even though these are usually considered equivalent.  Thus the refactor to avoid direct instantiation of `RatchetTreeNode`.